tests/functionaltests/Tests/NSURLConnectionTests.mm

```diff
@@ -19,6 +19,96 @@
 #import <atomic>
 #import <map>
 #import <random>
+#import <memory>
+
+@interface THRunLoopSpinner : NSObject {
+@public
+    std::shared_ptr<std::atomic<bool>> _shouldStop;
+    BOOL _threadStarted;
+    NSCondition* _threadStartedCondition;
+    NSRunLoop* _loop;
+    pthread_t _thread;
+}
+@property (nonatomic, readonly, retain) NSRunLoop* loop;
+@end
+
+@implementation THRunLoopSpinner
+static void* _helperThreadBody(void* context) {
+    __unsafe_unretained THRunLoopSpinner* unsafeHelper = (__bridge id)context;
+    @autoreleasepool {
+        std::shared_ptr<std::atomic<bool>> pShouldStop;
+        NSRunLoop* loop = nil;
+        { // scope guard for ownership of helper
+            __strong THRunLoopSpinner* helper = unsafeHelper;
+            pShouldStop =
+                helper->_shouldStop; // Capture the shared_ptr; the thread needs to read from it _after the helper is deallocated!_
+
+            loop = helper->_loop = [NSRunLoop currentRunLoop];
+
+            helper->_threadStarted = YES;
+            [helper->_threadStartedCondition lock];
+            [helper->_threadStartedCondition broadcast];
+            [helper->_threadStartedCondition unlock];
+        }
+
+        std::atomic<bool>& shouldStop = *pShouldStop;
+        while (!shouldStop) {
+            @autoreleasepool {
+                [loop runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
+            }
+        }
+    }
+    return nullptr;
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        _shouldStop.reset(new std::atomic<bool>(false));
+        _threadStartedCondition = [NSCondition new];
+
+        [_threadStartedCondition lock];
+
+        auto rc = pthread_create(&_thread, nullptr, &_helperThreadBody, (__bridge void*)self);
```

Comment:
Yep! An NSThread is just a pthread that's run through the NSThread APIs. NSRunLoop and NSThread work together to make sure that even untracked native threads get run loops.

Unfortunately, NSThread just can't be used for this anyway :)